File: GPU Programming/Flash Attention/FlashAttention_Chapter_2/online_softmax.py
"""Chapter 2: from 3-pass softmax to a one-pass softmax-weighted sum.

Pure-Python loops on purpose: the loop structure is the point, and it is what we
translate to C++ (and later CUDA). numpy is used only to check answers.
"""
import math
import random

import numpy as np

NEG_INF = float("-inf")


# --------------------------------------------------------------------------- #
# 1. Safe softmax: three passes over the input
# --------------------------------------------------------------------------- #
def softmax_3pass(x):
    m = NEG_INF
    for xi in x:                       # pass 1: maximum
        m = max(m, xi)
    d = 0.0
    for xi in x:                       # pass 2: normaliser
        d += math.exp(xi - m)
    return [math.exp(xi - m) / d for xi in x]   # pass 3: outputs


# --------------------------------------------------------------------------- #
# 2. Online normaliser: max and sum in ONE pass, then one pass for the outputs
# --------------------------------------------------------------------------- #
def push(m, d, x):
    """Fold one more score x into the running state (m, d).

    Invariant after the call:  m = max of everything seen so far,
                               d = sum over everything seen of exp(value - m).
    """
    m_new = max(m, x)
    if m_new == NEG_INF:               # everything seen so far is masked: nothing to accumulate
        return m, d
    d = d * math.exp(m - m_new) + math.exp(x - m_new)
    return m_new, d


def online_stats(x, trace=False):
    m, d = NEG_INF, 0.0
    for j, xj in enumerate(x):
        m_old = m
        m, d = push(m, d, xj)
        if trace:
            print(f"  j={j}  x={xj:5.2f}  m_old={m_old:5.2f}  m_new={m:5.2f}  "
                  f"rescale=exp(m_old-m_new)={math.exp(m_old - m):.4f}  d={d:.4f}")
    return m, d


def softmax_online(x):
    m, d = online_stats(x)
    if d == 0.0:                       # fully masked row: our convention is all zeros
        return [0.0] * len(x)
    return [math.exp(xi - m) / d for xi in x]


def softmax_online_unguarded(x):
    """The 'obvious' version, WITHOUT the masked-so-far guard. Kept to show the failure."""
    m, d = NEG_INF, 0.0
    with np.errstate(all="ignore"):
        for xi in x:
            m_new = max(m, xi)
            d = d * np.exp(m - m_new) + np.exp(xi - m_new)
            m = m_new
    return [float(np.exp(xi - m) / d) for xi in x]


# --------------------------------------------------------------------------- #
# 3. The merge operator: combine the states of two chunks of a row
# --------------------------------------------------------------------------- #
def merge(a, b):
    (ma, da), (mb, db) = a, b
    m = max(ma, mb)
    if m == NEG_INF:
        return (NEG_INF, 0.0)
    return (m, da * math.exp(ma - m) + db * math.exp(mb - m))


def stats_blocked(x, block):
    total = (NEG_INF, 0.0)
    for start in range(0, len(x), block):
        m, d = NEG_INF, 0.0
        for xi in x[start:start + block]:
            m, d = push(m, d, xi)
        total = merge(total, (m, d))
    return total


# --------------------------------------------------------------------------- #
# 4. One pass for what attention actually needs: sum_j softmax(s)_j * v_j
#    v_j may be a vector (numpy array); the same rescale applies to every component.
# --------------------------------------------------------------------------- #
def weighted_sum_online(s, V):
    m, d = NEG_INF, 0.0
    o = np.zeros_like(V[0], dtype=np.float64)
    for sj, vj in zip(s, V):
        m_new = max(m, sj)
        if m_new == NEG_INF:
            continue
        scale = math.exp(m - m_new)     # shrink everything accumulated so far
        w = math.exp(sj - m_new)        # weight of the new element
        d = d * scale + w
        o = o * scale + w * vj
        m = m_new
    return o / d if d > 0.0 else np.zeros_like(o)


def weighted_sum_reference(s, V):
    p = softmax_3pass(s)
    return sum(pj * vj for pj, vj in zip(p, V))


if __name__ == "__main__":
    np.set_printoptions(precision=6, suppress=True)
    rng = random.Random(0)

    print("== A. trace of the online normaliser on x = [1, 3, 2, 4] ==")
    m, d = online_stats([1.0, 3.0, 2.0, 4.0], trace=True)
    direct = sum(math.exp(v - 4.0) for v in [1.0, 3.0, 2.0, 4.0])
    print(f"  final (m, d) = ({m}, {d:.6f});  direct sum exp(x - 4) = {direct:.6f}")

    print("\n== B. online vs 3-pass vs numpy on awkward inputs (max abs diff) ==")
    N = 257
    families = {
        "random normal":        [rng.gauss(0, 3) for _ in range(N)],
        "offset +1000":         [1000 + rng.gauss(0, 3) for _ in range(N)],
        "increasing":           [float(i) * 0.1 for i in range(N)],
        "decreasing":           [-float(i) * 0.1 for i in range(N)],
        "constant":             [7.0] * N,
        "single element":       [3.5],
        "-inf in the middle":   [rng.gauss(0, 1) if i % 5 else NEG_INF for i in range(N)],
        "-inf at the START":    [NEG_INF] * 3 + [rng.gauss(0, 1) for _ in range(N - 3)],
    }
    for name, x in families.items():
        ref = np.exp(np.array(x) - np.max(x)); ref = ref / ref.sum()
        a, b = np.array(softmax_3pass(x)), np.array(softmax_online(x))
        print(f"  {name:<20} |3pass-ref|={np.abs(a - ref).max():.2e}  "
              f"|online-ref|={np.abs(b - ref).max():.2e}  sum(online)={b.sum():.6f}")

    print("\n== C. the guard: a row that starts with -inf ==")
    x = [NEG_INF, NEG_INF, 1.0, 2.0, 0.5]
    print("  unguarded:", np.array(softmax_online_unguarded(x)))
    print("  guarded  :", np.array(softmax_online(x)))
    print("  reference:", np.array(softmax_3pass(x[2:])), "(the last three entries)")
    print("  all -inf row, guarded:", softmax_online([NEG_INF] * 3), "(our convention: zeros)")

    print("\n== D. merging chunks: block size and merge order do not matter ==")
    x = [rng.gauss(0, 4) for _ in range(1000)]
    m0, d0 = online_stats(x)
    for block in (1, 2, 3, 7, 64, 500, 1000):
        m, d = stats_blocked(x, block)
        print(f"  block={block:>4}  m={m:.6f}  d={d:.10f}  |d - d_sequential|/d = {abs(d - d0) / d0:.2e}")
    # random binary-tree merge order
    parts = []
    for start in range(0, len(x), 10):
        m, d = NEG_INF, 0.0
        for xi in x[start:start + 10]:
            m, d = push(m, d, xi)
        parts.append((m, d))
    worst = 0.0
    for _ in range(200):
        pool = parts[:]
        rng.shuffle(pool)
        while len(pool) > 1:
            i, j = rng.sample(range(len(pool)), 2)
            a, b = pool[i], pool[j]
            pool = [p for k, p in enumerate(pool) if k not in (i, j)] + [merge(a, b)]
        worst = max(worst, abs(pool[0][1] - d0) / d0)
    print(f"  200 random merge trees over 100 chunks: worst relative error in d = {worst:.2e}")

    print("\n== E. how often does the maximum change? (rescale is only needed then) ==")
    for n in (16, 256, 4096):
        trials, changes = 200, 0
        for _ in range(trials):
            m = NEG_INF
            for _ in range(n):
                v = rng.gauss(0, 1)
                if v > m:
                    changes += 1
                    m = v
        harmonic = sum(1.0 / k for k in range(1, n + 1))
        print(f"  N={n:>5}: mean max-updates per row = {changes / trials:5.2f}   (harmonic number H_N = {harmonic:5.2f})")
    inc = [float(i) for i in range(256)]
    changes = sum(1 for i in range(1, 256) if inc[i] > max(inc[:i])) + 1
    print(f"  increasing input, N=256: max changes on {changes} of 256 elements (worst case)")

    print("\n== F. one pass for the weighted sum, vector values (N=64, d=8) ==")
    worst = 0.0
    for trial in range(200):
        s = [rng.gauss(0, 3) for _ in range(64)]
        if trial % 4 == 0:
            s[:5] = [NEG_INF] * 5          # masked start
        V = [np.array([rng.gauss(0, 1) for _ in range(8)]) for _ in range(64)]
        worst = max(worst, np.abs(weighted_sum_online(s, V) - weighted_sum_reference(s, V)).max())
    print(f"  worst |one-pass - reference| over 200 rows: {worst:.2e}")

File: GPU Programming/Flash Attention/FlashAttention_Chapter_2/online_softmax.cpp
// Chapter 2: online softmax in C++ (scalar, CPU only).
// Build:  g++ -std=c++17 -O2 -Wall -Wextra -o online_softmax online_softmax.cpp
// Run:    ./online_softmax
#include <algorithm>   // std::max, std::min
#include <cmath>       // expf, INFINITY, fabsf
#include <cstdint>     // uint32_t
#include <cstdio>      // printf
#include <vector>      // std::vector

// ---------------------------------------------------------------------------
// Deterministic pseudo-random numbers. The Python checker (check_cpp.py)
// reproduces the exact same sequence, so both languages see identical inputs.
// ---------------------------------------------------------------------------
void lcg_fill(std::vector<float>& x, uint32_t seed, float scale) {
    uint32_t state = seed;
    for (size_t i = 0; i < x.size(); ++i) {
        state = 1664525u * state + 1013904223u;                    // wraps modulo 2^32
        float u = static_cast<float>(state >> 8) / 16777216.0f;    // 24-bit integer / 2^24: exact
        x[i] = (2.0f * u - 1.0f) * scale;
    }
}

// ---------------------------------------------------------------------------
// 1. Safe softmax, three passes.
// ---------------------------------------------------------------------------
std::vector<float> softmax_3pass(const std::vector<float>& x) {
    const size_t n = x.size();

    float m = -INFINITY;                       // pass 1: maximum
    for (size_t i = 0; i < n; ++i) {
        m = std::max(m, x[i]);
    }

    float d = 0.0f;                            // pass 2: normaliser
    for (size_t i = 0; i < n; ++i) {
        d += expf(x[i] - m);
    }

    std::vector<float> y(n);                   // pass 3: outputs
    for (size_t i = 0; i < n; ++i) {
        y[i] = expf(x[i] - m) / d;
    }
    return y;
}

// ---------------------------------------------------------------------------
// 2. Online normaliser: the running state is a (max, sum) pair.
// ---------------------------------------------------------------------------
struct MD {
    float m;   // running maximum
    float d;   // running sum of exp(value - m)
};

// Fold one more score into the running state.
MD push(MD s, float x) {
    float m_new = std::max(s.m, x);
    if (m_new == -INFINITY) {                  // everything so far is masked
        return s;
    }
    return MD{m_new, s.d * expf(s.m - m_new) + expf(x - m_new)};
}

MD online_stats(const std::vector<float>& x) {
    MD s{-INFINITY, 0.0f};
    for (float xi : x) {                       // range-based for: like `for xi in x`
        s = push(s, xi);
    }
    return s;
}

std::vector<float> softmax_online(const std::vector<float>& x) {
    MD s = online_stats(x);
    std::vector<float> y(x.size(), 0.0f);      // all zeros: also our answer for a fully masked row
    if (s.d == 0.0f) {
        return y;
    }
    for (size_t i = 0; i < x.size(); ++i) {
        y[i] = expf(x[i] - s.m) / s.d;
    }
    return y;
}

// ---------------------------------------------------------------------------
// 3. Merge two partial states (what makes tiling and parallel reduction work).
// ---------------------------------------------------------------------------
MD merge(MD a, MD b) {
    float m = std::max(a.m, b.m);
    if (m == -INFINITY) {
        return MD{-INFINITY, 0.0f};
    }
    return MD{m, a.d * expf(a.m - m) + b.d * expf(b.m - m)};
}

MD stats_blocked(const std::vector<float>& x, size_t block) {
    MD total{-INFINITY, 0.0f};
    for (size_t start = 0; start < x.size(); start += block) {
        size_t end = std::min(start + block, x.size());
        MD part{-INFINITY, 0.0f};
        for (size_t i = start; i < end; ++i) {
            part = push(part, x[i]);
        }
        total = merge(total, part);
    }
    return total;
}

// ---------------------------------------------------------------------------
// 4. One pass for sum_j softmax(s)_j * v_j  (v_j is a single number here;
//    vector-valued v_j is the same update on every component: Chapter 3).
// ---------------------------------------------------------------------------
float weighted_average_online(const std::vector<float>& s, const std::vector<float>& v) {
    float m = -INFINITY;
    float d = 0.0f;
    float o = 0.0f;                            // running UNNORMALISED output
    for (size_t j = 0; j < s.size(); ++j) {
        float m_new = std::max(m, s[j]);
        if (m_new == -INFINITY) {
            continue;
        }
        float scale = expf(m - m_new);         // shrink everything accumulated so far
        float w = expf(s[j] - m_new);          // weight of the new element
        d = d * scale + w;
        o = o * scale + w * v[j];
        m = m_new;
    }
    return d == 0.0f ? 0.0f : o / d;           // normalise once, at the end
}

float weighted_average_reference(const std::vector<float>& s, const std::vector<float>& v) {
    std::vector<float> p = softmax_3pass(s);
    float acc = 0.0f;
    for (size_t j = 0; j < s.size(); ++j) {
        acc += p[j] * v[j];
    }
    return acc;
}

// ---------------------------------------------------------------------------
int main() {
    const size_t n = 16;
    std::vector<float> x(n);
    lcg_fill(x, 12345u, 4.0f);

    std::vector<float> y3 = softmax_3pass(x);
    std::vector<float> y2 = softmax_online(x);

    float max_diff = 0.0f;
    float sum = 0.0f;
    for (size_t i = 0; i < n; ++i) {
        max_diff = std::max(max_diff, fabsf(y3[i] - y2[i]));
        sum += y2[i];
        printf("y3 %zu %.7f\n", i, y3[i]);
        printf("y2 %zu %.7f\n", i, y2[i]);
    }
    printf("max |3pass - online| = %.3e   sum(online) = %.7f\n", max_diff, sum);

    MD ref = online_stats(x);
    for (size_t block : {1, 3, 5, 16}) {
        MD s = stats_blocked(x, block);
        printf("blocked block=%2zu  m=%.7f  d=%.7f  |d - d_seq| = %.2e\n",
               block, s.m, s.d, fabsf(s.d - ref.d));
    }

    std::vector<float> v(n);
    lcg_fill(v, 777u, 2.0f);
    float a = weighted_average_online(x, v);
    float b = weighted_average_reference(x, v);
    printf("weighted average: one-pass = %.7f  reference = %.7f\n", a, b);

    std::vector<float> masked = x;
    masked[0] = masked[1] = masked[2] = -INFINITY;     // a row that starts masked
    float c = weighted_average_online(masked, v);
    std::vector<float> tail_s(masked.begin() + 3, masked.end());
    std::vector<float> tail_v(v.begin() + 3, v.end());
    float e = weighted_average_reference(tail_s, tail_v);
    printf("masked start:     one-pass = %.7f  reference = %.7f\n", c, e);

    // A longer row: now the two algorithms round differently.
    std::vector<float> big(4096);
    lcg_fill(big, 1u, 8.0f);
    std::vector<float> b3 = softmax_3pass(big);
    std::vector<float> b2 = softmax_online(big);
    float big_diff = 0.0f;
    float big_sum3 = 0.0f;
    float big_sum2 = 0.0f;
    for (size_t i = 0; i < big.size(); ++i) {
        big_diff = std::max(big_diff, fabsf(b3[i] - b2[i]));
        big_sum3 += b3[i];
        big_sum2 += b2[i];
    }
    printf("n=4096: max |3pass - online| = %.3e   sum(3pass) = %.6f   sum(online) = %.6f\n",
           big_diff, big_sum3, big_sum2);
    return 0;
}
